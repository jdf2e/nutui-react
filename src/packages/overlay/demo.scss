.demo-overlay {
  min-height: 100vh;
}
.wrapper {
  display: flex;
  height: 100%;
  align-items: center;
  justify-content: center;
  .content {
    display: flex;
    width: 150px;
    height: 150px;
    background: #fff;
    border-radius: 8px;
    align-items: center;
    justify-content: center;
    color: red;
  }
}
