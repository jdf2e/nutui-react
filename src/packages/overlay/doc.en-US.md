# Overlay

## Intro

Create a mask layer that is typically used to prevent users from doing other things

## Install

```tsx
import { OverLay } from '@nutui/nutui-react';
```

## Demo

### Basic usage

:::demo

```tsx
import React, { useState } from "react";
import { Button, Overlay } from '@nutui/nutui-react';

const App = () => {
  const [visible, setVisible] = useState(false)
  const handleToggleShow = () => {
    setVisible(true)
  }
  const onClose = () => {
    setVisible(false)
  }
  return (
    <>
      <Button type="primary" onClick={handleToggleShow}>
        Displays the mask layer
      </Button>
      <Overlay
        visible={visible}
        onClick={onClose}
        zIndex={2020}
        afterShow={() => {
          console.log('afterShow')
        }}
      />
    </>
  )
}
export default App;
```

:::

### Mask style

:::demo

```tsx
import React, { useState } from "react";
import { Button, Overlay } from '@nutui/nutui-react';

const App = () => {
  const [visible, setVisible] = useState(false)
  const handleToggleShow = () => {
    setVisible(true)
  }
  const onClose = () => {
    setVisible(false)
  }
  return (
    <>
      <Button type="primary" onClick={handleToggleShow}>
        Mask style
      </Button>
      <Overlay
        visible={visible}
        onClick={onClose}
        zIndex={2000}
        style={{
          backgroundColor: 'rgba(0, 0, 0, .2)',
        }}
      />
    </>
  )
}
export default App;
```

:::

### Set animation time

:::demo

```tsx
import React, { useState } from "react";
import { Button, Overlay } from '@nutui/nutui-react';

const App = () => {
  const [visible, setVisible] = useState(false)
  const handleToggleShow = () => {
    setVisible(true)
  }
  const onClose = () => {
    setVisible(false)
  }
  return (
    <>
      <Button type="primary" onClick={handleToggleShow}>
        Set animation time
      </Button>
      <Overlay
        visible={visible}
        onClick={onClose}
        duration={2.5}
        afterShow={() => {
          console.log('afterShow')
        }}
        afterClose={() => {
          console.log('afterClose')
        }}
      />
    </>
  )
}
export default App;
```

:::

### Do not lock background scrolling

:::demo

```tsx
import React, { useState } from "react";
import { Button, Overlay } from '@nutui/nutui-react';

const App = () => {
  const [visible, setVisible] = useState(false)
  const handleToggleShow = () => {
    setVisible(true)
  }
  const onClose = () => {
    setVisible(false)
  }
  return (
    <>
      <Button type="primary" onClick={handleToggleShow}>
        Do not lock background scrolling
      </Button>
      <Overlay visible={visible} onClick={onClose} lockScroll={false} />
    </>
  )
}
export default App;
```

:::

### Nested content

:::demo

```tsx
import React, { useState } from "react";
import { Button, Overlay } from '@nutui/nutui-react';

const WrapperStyle = {
  display: 'flex',
  height: '100%',
  alignItems: 'center',
  justifyContent: 'center'
}
const ContentStyle = {
  display: 'flex',
  width: '150px',
  height: '150px',
  background: '#fff',
  borderRadius: '8px',
  alignItems: 'center',
  justifyContent: 'center',
  color: 'red'
}
const App = () => {
  const [visible, setVisible] = useState(false)
  const handleToggleShow = () => {
    setVisible(true)
  }
  const onClose = () => {
    setVisible(false)
  }
  return (
    <>
      <Button type="success" onClick={handleToggleShow}>
        Nested content
      </Button>
      <Overlay visible={visible} onClick={onClose}>
        <div className="wrapper" style={WrapperStyle}>
          <div className="content" style={ContentStyle}>Here is the main text</div>
        </div>
      </Overlay>
    </>
  )
}
export default App;
```

:::

### Click the mask not to close

:::demo

```tsx
import React, { useState } from "react";
import { Button, Overlay } from '@nutui/nutui-react';

const App = () => {
  const [visible, setVisible] = useState(false)
  const handleToggleShow = () => {
    setVisible(true)
  }
  const onClose = () => {
    setVisible(false)
  }
  return (
    <>
      <Button type="primary" onClick={handleToggleShow}>
        Click the mask not to close
      </Button>
      <Overlay visible={visible} closeOnOverlayClick={false}>
        <div className="wrapper">
          <div className="content" onClick={onClose}>here is the text</div>
        </div>
      </Overlay>
    </>
  )
}
export default App;
```

:::

## Overlay

### Props

| Property | Description | Type | Default |
| --- | --- | --- | --- |
| visible | Whether the current component is displayed | `boolean` | `false` |
| zIndex | Mask hierarchy | `number` | `2000` |
| duration | Animation duration, in seconds | `number` | `0.3` |
| lockScroll | Whether the background is locked | `boolean` | `true` |
| closeOnOverlayClick | Tap Mask off | `boolean` | `true` |
| onClick | Triggered when the button is clicked | `event: Event` | `-` |
| afterClose | Triggered after complete shutdown | `() => void` | `-` |
| afterShow | Trigger after complete display | `() => void` | `-` |

## Theming

### CSS Variables

The component provides the following CSS variables, which can be used to customize styles. Please refer to [ConfigProvider component](#/en-US/component/configprovider).

| Name | Description | Default Value |
| --- | --- | --- |
| \--nutui-overlay-bg-color | Overlay background color | `$gray7` |
| \--nutui-overlay-content-bg-color | Mask layer nested content background color | `$gray6` |
| \--nutui-overlay-content-color | Mask layer nested content font color | `$gray1` |