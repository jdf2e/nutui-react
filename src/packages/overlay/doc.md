# Overlay 遮罩层

### 介绍

创建一个遮罩层，通常用于阻止用户进行其他操作

### 安装


``` javascript
import { OverLay } from '@nutui/nutui-react';
```

## 代码演示

### 基础用法

```tsx
<Overlay visible={true} zindex={2000}></Overlay>
```

### 嵌套内容

```tsx
<nut-overlay visible={true} zIndex={2000}>
  <div className="wrapper">
    <div className="content">这里是正文</div>
  </div>
</nut-overlay>
```

## API

### Props

| 参数                   | 说明             | 类型           | 默认值 |
| ---------------------- | ---------------- | -------------- | ------ |
| visible                   | 当前组件是否显示 | Boolean        | `false`  |
| zIndex                | 遮罩层级         | Number | 2000   |
| duration               | 动画时长，单位秒 | Number | 0.3    |
| overlayClass          | 自定义遮罩类名   | String         | -      |
| overlayStyle          | 自定义遮罩样式   | CSSProperties  | -      |
| lockScroll          | 背景是否锁定   | Boolean  | `false`     |
| closeOnClickOverlay | 是否点击遮罩关闭 | Boolean        | `true`   |

### Events

| 事件名 | 说明       | 回调参数     |
| ------ | ---------- | ------------ |
| onClick  | 点击时触发 | event: Event |
