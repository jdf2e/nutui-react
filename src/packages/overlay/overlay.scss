.nut-theme-dark {
  .nut-overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: $dark7;
  }
  .overlay-fade-enter-active {
    .wrapper {
      .content {
        background-color: $dark2;
        color: $white;
      }
    }
  }
}

.nut-overlay {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: $overlay-bg-color;
}
.nut-overflow-hidden {
  overflow: hidden !important;
}
.overlay-fade-enter-active {
  animation: nut-fade-in;
}
.overlay-fade-leave-active {
  animation: nut-fade-out;
}
.first-render {
  display: none;
}
.hidden-render {
  display: none;
}
@keyframes nut-fade-in {
  0% {
    opacity: 0;
  }

  1% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

@keyframes nut-fade-out {
  0% {
    opacity: 1;
  }

  1% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
}
