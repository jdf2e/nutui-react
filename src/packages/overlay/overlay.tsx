import React, {
  FunctionComponent,
  MouseEventHandler,
  useEffect,
  useRef,
  useState,
} from 'react'
import classNames from 'classnames'
import bem from '@/utils/bem'

export interface OverlayProps {
  zIndex: number
  duration: number
  overlayClass: string
  overlayStyle: React.CSSProperties
  closeOnClickOverlay: boolean
  visible: boolean
  lockScroll: boolean
}
export const defaultOverlayProps = {
  zIndex: 1000,
  duration: 0.3,
  overlayClass: '',
  closeOnClickOverlay: true,
  visible: false,
  lockScroll: true,
  overlayStyle: {},
} as OverlayProps
export const Overlay: FunctionComponent<
  Partial<OverlayProps> & React.HTMLAttributes<HTMLDivElement>
> = (props) => {
  const {
    children,
    zIndex,
    duration,
    overlayClass,
    closeOnClickOverlay,
    visible,
    lockScroll,
    overlayStyle,
    ...rest
  } = {
    ...defaultOverlayProps,
    ...props,
  }
  const [show, setShow] = useState(visible)
  const renderRef = useRef(true)
  const intervalRef = useRef(0)

  useEffect(() => {
    visible && setShow(visible)
    lock()
  }, [visible])

  useEffect(() => {
    return () => {
      clearTimeout(intervalRef.current)
      document.body.classList.remove('nut-overflow-hidden')
    }
  }, [])

  const b = bem('overlay')

  const classes = classNames(
    {
      'overlay-fade-leave-active': !renderRef.current && !visible,
      'overlay-fade-enter-active': visible,
      'first-render': renderRef.current && !visible,
      'hidden-render': !visible,
    },
    overlayClass,
    b('')
  )

  const styles = {
    zIndex,
    animationDuration: `${props.duration}s`,
    ...overlayStyle,
  }

  const lock = () => {
    if (lockScroll && visible) {
      document.body.classList.add('nut-overflow-hidden')
    } else {
      document.body.classList.remove('nut-overflow-hidden')
    }
  }

  const handleClick: MouseEventHandler<HTMLDivElement> = (e) => {
    if (closeOnClickOverlay) {
      props.onClick && props.onClick(e)
      renderRef.current = false
      const id = setTimeout(() => {
        setShow(!visible)
      }, duration * 1000 * 0.8)
      intervalRef.current = id
    }
  }

  return (
    <>
      <div className={classes} style={styles} {...rest} onClick={handleClick}>
        {children}
      </div>
    </>
  )
}

Overlay.defaultProps = defaultOverlayProps
Overlay.displayName = 'NutOverlay'
