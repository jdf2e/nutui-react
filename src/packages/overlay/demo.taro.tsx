import React, { useState } from 'react'
import Taro from '@tarojs/taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import { Button, Cell, Overlay } from '@/packages/nutui.react.taro'
import '@/packages/overlay/demo.scss'
import Header from '@/sites/components/header'

interface T {
  '84aa6bce': string
  '2a9e4928': string
  duration: string
  lockscroll: string
  abbf9359: string
  ec0d7acf: string
  ce1e18a2: string
  closeClickLay: string
}
const OverlayDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      '84aa6bce': '基础用法',
      '2a9e4928': '显示遮罩层',
      duration: '设置动画时间',
      lockscroll: '不锁定背景滚动',
      abbf9359: '自定义遮罩样式',
      ec0d7acf: '嵌套内容',
      ce1e18a2: '这里是正文',
      closeClickLay: '点击遮罩不关闭',
    },
    'zh-TW': {
      '84aa6bce': '基礎用法',
      '2a9e4928': '顯示遮罩層',
      duration: '設置動畫時間',
      lockscroll: '不鎖定背景滾動',
      abbf9359: '自定義遮罩樣式',
      ec0d7acf: '嵌套內容',
      ce1e18a2: '這裡是正文',
      closeClickLay: '點擊遮罩不關閉',
    },
    'en-US': {
      '84aa6bce': 'Basic usage',
      '2a9e4928': 'Show mask layer',
      duration: 'Set animation time',
      lockscroll: 'Donot lock background scrolling',
      abbf9359: 'Custom mask style',
      ec0d7acf: 'nested content',
      ce1e18a2: 'here is the text',
      closeClickLay: 'Click the mask not to close',
    },
  })

  const [visible, setVisible] = useState(false)
  const [visible2, setVisible2] = useState(false)
  const [visible3, setVisible3] = useState(false)
  const [visible4, setVisible4] = useState(false)
  const [visible5, setVisible5] = useState(false)
  const [visible6, setVisible6] = useState(false)

  const handleToggleShow = () => {
    setVisible(true)
  }
  const handleToggleShow2 = () => {
    setVisible2(true)
  }
  const handleToggleShow3 = () => {
    console.log('visible3', visible3)
    setVisible3(true)
  }
  const handleToggleShow4 = () => {
    setVisible4(true)
  }
  const handleToggleShow5 = () => {
    setVisible5(true)
  }
  const handleToggleShow6 = () => {
    setVisible6(true)
  }

  const onClose = () => {
    setVisible(false)
  }
  const onClose2 = () => {
    setVisible2(false)
  }
  const onClose3 = () => {
    setVisible3(false)
  }
  const onClose4 = () => {
    setVisible4(false)
  }
  const onClose5 = () => {
    setVisible5(false)
  }
  const onClose6 = () => {
    setVisible6(false)
  }
  return (
    <>
      <Header />
      <div
        className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''} demo-overlay`}
      >
        <h2>{translated['84aa6bce']}</h2>
        <Cell>
          <Button type="primary" onClick={handleToggleShow}>
            {translated['2a9e4928']}
          </Button>
          <Overlay
            visible={visible}
            onClick={onClose}
            style={{ '--nutui-overlay-zIndex': 2020 }}
            afterShow={() => {
              console.log('afterShow')
            }}
          />
        </Cell>

        <h2>{translated.abbf9359}</h2>
        <Cell>
          <Button type="primary" onClick={handleToggleShow2}>
            {translated.abbf9359}
          </Button>
          <Overlay
            visible={visible2}
            onClick={onClose2}
            style={{
              backgroundColor: 'rgba(0, 0, 0, .4)',
              '--nutui-overlay-zIndex': 2000,
            }}
          />
        </Cell>

        <h2>{translated.duration}</h2>
        <Cell>
          <Button type="primary" onClick={handleToggleShow3}>
            {translated.duration}
          </Button>
          <Overlay
            visible={visible3}
            onClick={onClose3}
            style={{ '--nutui-overlay-animation-duration': '2.5s' }}
            duration={2500}
            afterShow={() => {
              console.log('afterShow')
            }}
            afterClose={() => {
              console.log('afterClose')
            }}
          />
        </Cell>

        <h2>{translated.lockscroll}</h2>
        <Cell>
          <Button type="primary" onClick={handleToggleShow4}>
            {translated.lockscroll}
          </Button>
          <Overlay visible={visible4} lockScroll={false} onClick={onClose4} />
        </Cell>

        <h2>{translated.ec0d7acf}</h2>
        <Cell>
          <Button type="success" onClick={handleToggleShow5}>
            {translated.ec0d7acf}
          </Button>
          <Overlay visible={visible5} onClick={onClose5}>
            <div className="wrapper">
              <div className="content">{translated.ce1e18a2}</div>
            </div>
          </Overlay>
        </Cell>

        <h2>{translated.closeClickLay}</h2>
        <Cell>
          <Button type="primary" onClick={handleToggleShow6}>
            {translated.closeClickLay}
          </Button>
          <Overlay visible={visible6} closeOnOverlayClick={false}>
            <div className="wrapper">
              <div className="content" onClick={onClose6}>
                {translated.ce1e18a2}
              </div>
            </div>
          </Overlay>
        </Cell>
      </div>
    </>
  )
}

export default OverlayDemo
