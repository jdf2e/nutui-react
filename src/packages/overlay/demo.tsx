import React, { useState } from 'react'
import { useTranslate } from '../../sites/assets/locale'
import { Overlay } from './overlay'
import Cell from '@/packages/cell'
import Button from '@/packages/button'
import './demo.scss'

interface T {
  '84aa6bce': string
  '2a9e4928': string
  abbf9359: string
  ec0d7acf: string
  ce1e18a2: string
}
const OverlayDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      '84aa6bce': '基础用法',
      '2a9e4928': '显示遮罩层',
      abbf9359: '遮罩样式',
      ec0d7acf: '嵌套内容',
      ce1e18a2: '这里是正文',
    },
    'zh-TW': {
      '84aa6bce': '基礎用法',
      '2a9e4928': '顯示遮罩層',
      abbf9359: '遮罩樣式',
      ec0d7acf: '嵌套內容',
      ce1e18a2: '這裡是正文',
    },
    'en-US': {
      '84aa6bce': 'Basic usage',
      '2a9e4928': 'show mask layer',
      abbf9359: 'mask style',
      ec0d7acf: 'nested content',
      ce1e18a2: 'here is the text',
    },
  })

  const [visible, setVisible] = useState(false)
  const [visible2, setVisible2] = useState(false)
  const [visible3, setVisible3] = useState(false)

  const handleToggleShow = () => {
    setVisible(true)
  }
  const handleToggleShow2 = () => {
    setVisible2(true)
  }
  const handleToggleShow3 = () => {
    setVisible3(true)
  }

  const onClose = () => {
    setVisible(false)
  }
  const onClose2 = () => {
    setVisible2(false)
  }
  const onClose3 = () => {
    setVisible3(false)
  }
  return (
    <>
      <div className="demo demo-overlay">
        <h2>{translated['84aa6bce']}</h2>
        <Cell>
          <Button type="primary" onClick={handleToggleShow}>
            {translated['2a9e4928']}
          </Button>
          <Overlay visible={visible} onClick={onClose} />
        </Cell>
        <h2>{translated.abbf9359}</h2>
        <Cell>
          <Button type="primary" onClick={handleToggleShow2}>
            {translated['2a9e4928']}
          </Button>
          <Overlay
            visible={visible2}
            onClick={onClose2}
            overlayStyle={{
              backgroundColor: '#38333333',
            }}
          />
        </Cell>
        <h2>{translated.ec0d7acf}</h2>
        <Cell>
          <Button type="success" onClick={handleToggleShow3}>
            {translated.ec0d7acf}
          </Button>
          <Overlay visible={visible3} onClick={onClose3}>
            <div className="wrapper">
              <div className="content">{translated.ce1e18a2}</div>
            </div>
          </Overlay>
        </Cell>
      </div>
    </>
  )
}

export default OverlayDemo
