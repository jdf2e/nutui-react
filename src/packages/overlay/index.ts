import { Overlay } from './overlay'

export type { OverlayProps } from './overlay'
export default Overlay
