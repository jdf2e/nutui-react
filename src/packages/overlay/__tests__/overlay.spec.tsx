import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import { Overlay } from '../overlay'

test('should change z-index when using z-index prop', () => {
  const { getByTestId } = render(
    <Overlay data-testid="overlay-zindex" visible zIndex={99} />
  )
  expect(getByTestId('overlay-zindex').style.zIndex).toBe('99')
})

test('prop close-on-click-overlay test', () => {
  const onClose = jest.fn()
  const { getByTestId } = render(
    <Overlay
      data-testid="overlay-closeFalsy"
      visible
      closeOnOverlayClick={false}
      onClick={onClose}
    />
  )
  fireEvent.click(getByTestId('overlay-closeFalsy'))
  expect(getByTestId('overlay-closeFalsy').style.display).toEqual('')
})

test('event click test', () => {
  const onClose = jest.fn()
  const { getByTestId } = render(
    <Overlay
      data-testid="overlay-click"
      visible
      closeOnOverlayClick
      onClick={onClose}
    />
  )
  fireEvent.click(getByTestId('overlay-click'))
  expect(onClose).toBeCalled()
})

test('content test', () => {
  const { container } = render(
    <Overlay visible>
      <div className="wrapper">
        <div className="content">这里是正文</div>
      </div>
    </Overlay>
  )
  expect(container).toMatchSnapshot()
})
