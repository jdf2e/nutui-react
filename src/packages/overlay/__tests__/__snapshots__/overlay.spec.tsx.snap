// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`content test 1`] = `
<div>
  <div
    class="nut-overlay"
  >
    <div
      class="wrapper"
    >
      <div
        class="content"
      >
        这里是正文
      </div>
    </div>
  </div>
</div>
`;
